src/txadmission.cpp

```diff
@@ -363,9 +343,36 @@ void ThreadTxAdmission()
                         LOG(MEMPOOL, "rejected orphan as likely contains old sighash");
                     }
                 }
-                else // If the problem wasn't that the tx is an orphan, then uncache the inputs since we likely won't
-                // need them again.
+                else
                 {
+                    recentRejects.insert(tx->GetHash());
+
+                    if (txd.whitelisted && GetBoolArg("-whitelistforcerelay", DEFAULT_WHITELISTFORCERELAY))
+                    {
+                        // Always relay transactions received from whitelisted peers, even
+                        // if they were already in the mempool or rejected from it due
+                        // to policy, allowing the node to function as a gateway for
+                        // nodes hidden behind it.
+                        //
+                        // Never relay transactions that we would assign a non-zero DoS
+                        // score for, as we expect peers to do the same with us in that
+                        // case.
+                        int nDoS = 0;
+                        if (!state.IsInvalid(nDoS) || nDoS == 0)
+                        {
+                            LOGA("Force relaying tx %s from whitelisted peer=%s\n", tx->GetHash().ToString(),
+                                txd.nodeName);
+                            RelayTransaction(tx);
+                        }
+                        else
+                        {
+                            LOGA("Not relaying invalid transaction %s from whitelisted peer=%d (%s)\n",
+                                tx->GetHash().ToString(), txd.nodeName, FormatStateMessage(state));
```

Comment:
If I am not mistaken, this should be `[...] peer=%s [...]` as `txd.nodeName` is a `std::string`.